src/fennel.fnl

```diff
@@ -0,0 +1,110 @@
+;; SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+(local utils (require :fennel.utils))
+(local parser (require :fennel.parser))
+(local compiler (require :fennel.compiler))
+(local specials (require :fennel.specials))
+(local repl (require :fennel.repl))
+
+(fn eval [str options ...]
+  ;; eval and dofile are considered "live" entry points, so we can assume
+  ;; that the globals available at compile time are a reasonable allowed list
+  ;; UNLESS there's a metatable on env, in which case we can't assume that
+  ;; pairs will return all the effective globals; for instance openresty
+  ;; sets up _G in such a way that all the globals are available thru
+  ;; the __index meta method, but as far as pairs is concerned it's empty.
+  (let [opts (utils.copy options)
+        _ (when (and (= opts.allowedGlobals nil)
+                     (not (getmetatable opts.env)))
+            (set opts.allowedGlobals (specials.currentGlobalNames opts.env)))
+        env (and opts.env (specials.wrapEnv opts.env))
+        lua-source (compiler.compileString str opts)
+        loader (specials.loadCode lua-source env
+                                  (if opts.filename
+                                      (.. "@" opts.filename) str))]
+    (set opts.filename nil)
+    (loader ...)))
+
+(fn dofile* [filename options ...]
+  (let [opts (utils.copy options)
+        f (assert (io.open filename :rb))
+        source (f:read :*all)]
+    (f:close)
+    (set opts.filename filename)
+    (eval source opts ...)))
+
+;; The public API module we export:
+(local mod {
+            :list utils.list
+            :sym utils.sym
+            :varg utils.varg
+            :path utils.path
+
+            :parser parser.parser
+            :granulate parser.grandulate
+            :stringStream parser.stringStream
+
+            :compile compiler.compile
+            :compileString compiler.compileString
+            :compileStream compiler.compileStream
+            :compile1 compiler.compile1
+            :traceback compiler.traceback
+            :mangle compiler.globalMangling
+            :unmangle compiler.globalUnmangling
+            :metadata compiler.metadata
+            :scope compiler.makeScope
+            :gensym compiler.gensym
+
+            :loadCode specials.loadCode
+            :macroLoaded specials.macroLoaded
+            :searchModule specials.searchModule
+            :makeSearcher specials.makeSearcher
+            :make_searcher specials.makeSearcher ; backwards-compatibility alias
+            :searcher (specials.makeSearcher)
+            :doc specials.doc
+
+            :eval eval
+            :dofile dofile*
+            :version "0.5.0-dev"
+
+            :repl repl
+            })
+
+;; This is bad; we have a circular dependency between the specials section and
+;; the evaluation section due to require-macros/import-macros, etc. For now
+;; stash it in the utils table, but we should untangle it
+(set utils.fennelModule mod)
```

Comment:
What if we make some things lazy-loadable by setting a `__newindex` metamethod on modules that lazy-loads the circular things? That way it doesn't try to invoke the code with the circular dep until the other dep has already loaded.

That said, I think we could merge this and play with that in a subsequent PR.

Reply:
I'm not sure I follow but yeah, if you want to take a shot at it after the merge that'd be great. I don't really like the way this looks right now.